File: LiDAR2BIM-Registration/preprocess/clustering_utils.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def union(self, a, b):
        root_a = self.find(a)
        root_b = self.find(b)
        if root_a != root_b:
            if self.rank[root_a] > self.rank[root_b]:
                self.parent[root_b] = root_a
            else:
                self.parent[root_a] = root_b
                if self.rank[root_a] == self.rank[root_b]:
                    self.rank[root_b] += 1

    def find(self, a):
        if self.parent[a] != a:
            self.parent[a] = self.find(self.parent[a])
        return self.parent[a]

File: LiDAR2BIM-Registration/src/utils/clustering_utils.cpp
#include "utils/clustering_utils.h"
UnionFind::UnionFind(int n) {
	parent.resize(n);
	rank.resize(n, 0);
	for (int i = 0; i < n; ++i) {
		parent[i] = i;
	}
}

void UnionFind::unionSets(int a, int b) {
	int rootA = find(a);
	int rootB = find(b);
	if (rootA != rootB) {
		if (rank[rootA] > rank[rootB]) {
			parent[rootB] = rootA;
		} else {
			parent[rootA] = rootB;
			if (rank[rootA] == rank[rootB]) {
				rank[rootB] += 1;
			}
		}
	}
}

int UnionFind::find(int a) {
	if (parent[a] != a) {
		parent[a] = find(parent[a]);
	}
	return parent[a];
}